segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/tables/IteratorItem.java

```diff
@@ -0,0 +1,33 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.contracts.tables;
+
+import io.pravega.common.util.ArrayView;
+import io.pravega.common.util.AsyncIterator;
+import java.util.Collection;
+
+/**
+ * Defines an iteration result that is returned by the {@link AsyncIterator} when invoking {@link TableStore#keyIterator}
+ * or {@link TableStore#entryIterator}.
+ */
+public interface IteratorItem<T> {
+    /**
+     * Gets an array that represents the current state of the iteration. This value can be used to to reinvoke
+     * {@link TableStore#keyIterator} or {@link TableStore#entryIterator} if a previous iteration has been interrupted.
+     * @return An {@link ArrayView} containing the serialized state.
+     */
+    ArrayView getState();
+
+    /**
+     * Gets a Collection of items that are contained in this instance. The items in this list are not necessarily related
+     * to each other, nor are they guaranteed to be in any particular order.
+     */
+    Collection<T> getEntries();
```

Comment:
The reason I chose a `Collection` here vs just the individual type is because I can only iterate over whole `TableBucket`s and not inside them - that is, a `TableBucket` is my iteration granularity (each `IteratorItem` refers to exactly one `TableBucket`). Should there be multiple table keys in the same bucket (due to collisions), I will return all of them (`TableBucketReader.findAll`). Most of the time this will contain just one item.

The processor of this interface will need to either include all the items in this collection or none at all. The `IteratorItem.getState` returns the Key Hash of the `TableBucket` which it refers to. 

Passing that state again to `keyIterator/entryIterator` will resume the iteration from the next `TableBucket` in the sequence.

Reply:
My point is not so much about the internal processing and how the elements are obtained, but the contract. If the interface is such that `getNext()` can return a collection of arbitrary size, then it is possible that an implementation will return the whole table in a single shot, which defeats the purpose of the iterator.

I'm trying to determine whether it makes sense to constrain the behavior in a meaningful manner. The alternatives are not great, though. If we explain in the contract what we do right now, then we are binding the interface to that one implementation. If we bound the size of a collection we return, then it is not realistic because there is no meaningful way of computing such a bound.

I don't have any better option, really, but I wanted to raise the point to see if you can think of anything else.

Reply:
In reality this will contain one item in the vast majority of cases. In very large tables (>100M entries), a very small number will contain 2 items. In extremely large tables (> 1 billion entries, assuming we have the capacity to store that much), a slightly larger number will contain 2 items and some may contain 3.

Unless we have a `KeyHasher` set to always return a constant value, or bucketize in a rather small number of Table Buckets - all of which are set in the code and are not user-configurable, it will be very hard for this collection to contain more than 2-3 items.